### PR TITLE
Improve configuration and resource use of `MemoryManager` and `DiskManager`

### PR DESCRIPTION
**Pull request by alamb**
# Which issue does this PR close?

Resolves https://github.com/apache/arrow-datafusion/issues/1636

 # Rationale for this change
Creation of an `RuntimeEnv` creates a temporary directory, even when it is not used. This is unecessary per-query overhead

Also, since there is no way to share `MemoryManger` and `DiskManager`s across `RuntimeEnv's` it is not possible to get  a "global" view of memory / disk use across multiple queries running concurrently.

Also, certain invalid memory configuration values will panic, rather than error when set incorrectly

# What changes are included in this PR?
Changes:
1. Add `DiskManagerConfig` and `MemoryManagerConfig` for configuring how disk and memory are managed
2. Adds proper error checking (rather than asserts/ errors) to ensure reasonable values for memory config
2. Allow re-using existing `MemoryManager` and `DiskManager` rather than always creating them new

# Are there any user-facing changes?
Will be a change to anyone who was using the RuntimeConfig but since it was introduced recently I don't think that is very many

### CODE REVIEW

File: datafusion/src/execution/context.rs [hunk 1]
```diff
@@ -1057,6 +1062,42 @@ impl ExecutionConfig {
         self.runtime = config;
         self
     }
+
+    /// Use an an existing [MemoryManager]
```
Comment (alamb) [inline datafusion/src/execution/context.rs:1066]: Here is the new public API

File: datafusion/src/execution/memory_manager.rs [hunk 2]
```diff
@@ -463,26 +546,53 @@ mod tests {
...
         assert_eq!(requester1.get_spills(), 1);
         assert_eq!(requester1.mem_used(), 10);
 
         assert_eq!(*runtime.memory_manager.requesters_total.lock().unwrap(), 40);
+    }
 
-        Ok(())
+
+    #[tokio::test]
+    #[should_panic(expected = "invalid max_memory. Expected greater than 0, got 0")]
```
Comment (alamb) [inline datafusion/src/execution/memory_manager.rs:572]: previously such config errors would `panic`

File: datafusion/src/execution/context.rs [hunk 3]
```diff
@@ -1245,6 +1286,31 @@ mod tests {
...
+        // configure with same memory / disk manager
+        let memory_manager = ctx1.runtime_env().memory_manager.clone();
+        let disk_manager = ctx1.runtime_env().disk_manager.clone();
+        let config = ExecutionConfig::new()
+            .with_existing_memory_manager(memory_manager.clone())
+            .with_existing_disk_manager(disk_manager.clone());
+
+        let ctx2 = ExecutionContext::with_config(config);
+
+        assert!(std::ptr::eq(Arc::as_ptr(&memory_manager), Arc::as_ptr(&ctx1.runtime_env().memory_manager)));
```
Comment (alamb) [inline datafusion/src/execution/context.rs:1304]: Here is the tests for the primary usecase we have in IOx -- sharing `DiskManager` and `MemoryManager`s across plans

**Comment by alamb** [comment]
FYI @tustvold  and @yjshen

**Review by yjshen — APPROVED** [review]
Great config design! I really like the introduced enum configs.

**Comment by alamb** [comment]
Thanks @yjshen  -- I think since this basically only affects you and I for the time being I will update this PR and merge it in

**Pull request merged**